presto-main/src/main/java/com/facebook/presto/operator/aggregation/AggregationCompiler.java

```diff
@@ -48,98 +58,70 @@
                     for (String name : getNames(outputFunction, aggregationAnnotation)) {
-                        List<Type> inputTypes = getInputTypes(inputFunction);
-                        Type outputType = AggregationUtils.getOutputType(outputFunction, stateSerializer, typeManager);
-
-                        AggregationMetadata metadata;
-                        try {
-                            MethodHandle inputHandle = lookup().unreflect(inputFunction);
-                            MethodHandle intermediateInputHandle = intermediateInputFunction == null ? null : lookup().unreflect(intermediateInputFunction);
-                            MethodHandle combineHandle = combineFunction == null ? null : lookup().unreflect(combineFunction);
-                            MethodHandle outputHandle = outputFunction == null ? null : lookup().unreflect(outputFunction);
-                            metadata = new AggregationMetadata(
-                                    generateAggregationName(name, outputType.getTypeSignature(), inputTypes.stream().map(Type::getTypeSignature).collect(toImmutableList())),
-                                    getParameterMetadata(inputFunction, aggregationAnnotation.approximate()),
-                                    inputHandle,
-                                    getParameterMetadata(intermediateInputFunction, false),
-                                    intermediateInputHandle,
-                                    combineHandle,
-                                    outputHandle,
-                                    stateClass,
-                                    stateSerializer,
-                                    stateFactory,
-                                    outputType,
-                                    aggregationAnnotation.approximate());
-                        }
-                        catch (IllegalAccessException e) {
-                            throw Throwables.propagate(e);
-                        }
-
-                        AccumulatorFactoryBinder factory = new LazyAccumulatorFactoryBinder(metadata, classLoader);
-                        builder.add(new InternalAggregationFunction(name, inputTypes, intermediateType, outputType, aggregationAnnotation.decomposable(), aggregationAnnotation.approximate(), factory));
+                        List<TypeSignature> inputTypes = getInputTypesSignatures(inputFunction);
+                        TypeSignature outputType = AggregationUtils.getOutputTypeSignature(outputFunction);
+
+                        builder.add(
+                                new BindableAggregationFunction(
+                                        new Signature(
+                                                name,
+                                                aggregationAnnotation.approximate() ? FunctionKind.APPROXIMATE_AGGREGATE : FunctionKind.AGGREGATE,
+                                                ImmutableList.of(), // TODO parse constrains from annotations
```

Comment:
Lets implement this. Also, can you add a commit that ports something like `MaxAggregationFunction` to be annotation based to make sure this new `BindableAggregationFunction` can support everything that's needed for it?